test/e2e/storage/drivers/proxy/portproxy.go

```diff
@@ -85,53 +93,50 @@ func Listen(ctx context.Context, clientset kubernetes.Interface, restConfig *res
 	prefix := fmt.Sprintf("port forwarding for %s", addr)
 	ctx, cancel := context.WithCancel(ctx)
 	l := &listener{
-		connections: make(chan *connection),
-		ctx:         ctx,
-		cancel:      cancel,
-		addr:        addr,
+		ctx:    ctx,
+		cancel: cancel,
+		addr:   addr,
 	}
 
-	var connectionsCreated, connectionsClosed int32
+	var connectionsCreated int
 
 	runForwarding := func() {
 		klog.V(2).Infof("%s: starting connection polling", prefix)
 		defer klog.V(2).Infof("%s: connection polling ended", prefix)
 
-		// This delay determines how quickly we notice when someone has
-		// connected inside the cluster. With socat, we cannot make this too small
-		// because otherwise we get many rejected connections. With the mock
-		// driver as proxy that doesn't happen as long as we don't
-		// ask for too many concurrent connections because the mock driver
-		// keeps the listening port open at all times and the Linux
-		// kernel automatically accepts our connection requests.
-		tryConnect := time.NewTicker(100 * time.Millisecond)
+		tryConnect := time.NewTicker(connectionPollInterval)
 		defer tryConnect.Stop()
 		for {
 			select {
 			case <-ctx.Done():
 				return
 			case <-tryConnect.C:
-				currentClosed := atomic.LoadInt32(&connectionsClosed)
-				openConnections := connectionsCreated - currentClosed
-				if openConnections >= maxConcurrentConnections {
-					break
-				}
-
-				klog.V(5).Infof("%s: trying to create a new connection #%d, %d open", prefix, connectionsCreated, openConnections)
-				stream, err := dial(ctx, fmt.Sprintf("%s #%d", prefix, connectionsCreated), dialer, addr.Port)
-				if err != nil {
-					klog.Errorf("%s: no connection: %v", prefix, err)
-					break
-				}
-				// Make the connection available to Accept below.
-				klog.V(5).Infof("%s: created a new connection #%d", prefix, connectionsCreated)
-				l.connections <- &connection{
-					stream:  stream,
-					addr:    addr,
-					counter: connectionsCreated,
-					closed:  &connectionsClosed,
-				}
-				connectionsCreated++
+				func() {
+					l.mutex.Lock()
+					defer l.mutex.Unlock()
+
+					for i, c := range l.connections {
+						if c == nil {
+							klog.V(5).Infof("%s: trying to create a new connection #%d", prefix, connectionsCreated)
+							stream, err := dial(ctx, fmt.Sprintf("%s #%d", prefix, connectionsCreated), dialer, addr.Port)
+							if err != nil {
+								klog.Errorf("%s: no connection: %v", prefix, err)
+								return
+							}
+							// Make the connection available to Accept below.
+							klog.V(5).Infof("%s: created a new connection #%d", prefix, connectionsCreated)
+							c := &connection{
```

Comment:
Won't this create a busy loop? Before this PR, there was a channel that would block until `Accept()` is called, now it seems `runForwarding()` calls `dial()` every `connectionPollInterval`, creating new connections / streams. Will `dial()` block?

Reply:
> Won't this create a busy loop?

It will poll at a regular, small interval. That's less efficient than the previous approach with the channel (small delay between "new connection available" and "Accept returns"; periodic wakeups), but should be good enough.

Reply:
`dial` only gets called when there are less than `maxConcurrentConnections` active connections, so this won't create connections continuously.

Reply:
Is goal of the code to really create 10 connections and then wait for 1 of them to get Accepted? I can see it works, but I find it odd.

Reply:
Maybe a comment there would help.

Reply:
The old code did the same: establishing connections is a lengthy process and was prone to timeouts. To avoid that, the proxy establishes up to 10 connections that are ready to carry bytes back and forth immediately when something connects to the remote listen socket. I thought that was explained somewhere, but it isn't - will add it.

Reply:
Done as comment for maxConcurrentConnections.